File: agent/reviews.py
#!/usr/bin/env python3
"""Reviews: a reading of a branch, handed to the session as a file.

A person reads a branch in the panel and writes on its lines. What comes of it
reaches the session as a file rather than as text in its terminal: the session
is told where the reading lies and opens it when it gets to it. A file is also
the cheaper thing to extend — a field added here is a field an older reader
steps over, while a line of text has to be parsed by whoever reads it.

The shelf is written from one side only. The panel runs in a container with the
state directory mounted read-only, so it cannot lay the file down itself: it
hands the reading over the socket here, and the file is named after the reading
on this side. That is why the name is checked and never repaired — a name that
had to be repaired is not a name the panel minted, and this is the one place
where reading a branch turns into writing to disk.

Nothing here renders anything. The reading is data: the quote of a line is kept
as the file reads, and what to make of it is the reader's business.
"""

import datetime
import json
import os
import re
import socket
import threading
import time

import paths

# The socket stands in the shelf directory rather than in /run, where the
# sockets of the sessions are: this one is dialled from the container, and the
# state directory is the only thing mounted there. Only *.json is ever read
# back as a reading, so a socket beside them is not one of them.
SOCKET_DIR = os.environ.get("AACP_REVIEW_DIR", paths.state("reviews"))
SOCKET_NAME = "review.sock"

STORE = os.environ.get("AACP_REVIEW_STORE", paths.state("reviews"))

# What a reading may hold. The panel holds a reading to the same numbers before
# it saves one, and they are repeated here because a file on this socket does
# not have to have come from the panel.
MAX_NOTES = 200
MAX_TEXT = 4000
MAX_QUOTE = 2000

MAX_ID = 120
MAX_NOTE_ID = 64
MAX_LINE = 400

# Two hundred notes at their ceilings, in an alphabet where a character costs
# four bytes, with the json around them. A request past this is not a reading.
MAX_REQUEST = 8 * 1024 * 1024

# How much of this the machine keeps.
#
# The age of the file and the age of the sending are two different rules on
# purpose. What a reading says about itself can be absent or a lie — a stamp
# set in the future would sit on the shelf for ever — so a coarse rule on the
# file itself stands behind the one that reads it.
MAX_REVIEWS = 64
MAX_STORE = 32 * 1024 * 1024
MAX_AGE = 90 * 24 * 3600
SENT_AGE = 30 * 24 * 3600

STAMP = "%Y-%m-%dT%H:%M:%SZ"

# The name of a reading as the panel mints it, and nothing else: it becomes the
# name of a file, and every character that is not this one is a way out of the
# shelf.
NAME = re.compile(r"^[A-Za-z0-9_-]{1,%d}$" % MAX_ID)


class Refused(ValueError):
    """The reading was not taken, and the reason is for the panel to show."""


def _line(value, limit=MAX_LINE):
    """One line: whitespace collapsed, cut to the ceiling."""
    return " ".join(str(value or "").split())[:limit]


def _body(value):
    """What somebody wrote: line breaks kept, ends trimmed."""
    return str(value or "").replace("\r\n", "\n").replace("\r", "\n").strip()


def _quote(value):
    """The line a note stands on, kept as the file reads it.

    Leading whitespace is the indentation of the line and is what tells two
    otherwise identical lines apart, so nothing is collapsed here.
    """
    return str(value or "").replace("\r\n", "\n").replace("\r", "\n").rstrip("\n")


def _count(value):
    """Returns a line number, or zero when what came is not one."""
    try:
        return int(value)
    except (TypeError, ValueError, OverflowError):
        return 0


def seconds(stamp):
    """Returns a stamp in seconds, or zero when it cannot be read."""
    if not isinstance(stamp, str) or not stamp:
        return 0
    try:
        when = datetime.datetime.fromisoformat(stamp.replace("Z", "+00:00"))
    except ValueError:
        return 0
    if when.tzinfo is None:
        when = when.replace(tzinfo=datetime.timezone.utc)
    return when.timestamp()


def _stamp(value):
    """Returns a time in the one shape the shelf compares, or an empty string.

    A time arrives as the panel wrote it, which is however Go prints one: with
    fractions, and with an offset that is not always Z. It is kept in one shape
    so that the sweep does not have to know about the others.
    """
    when = seconds(_line(value, 64))
    if not when:
        return ""
    return time.strftime(STAMP, time.gmtime(when))


def name(value):
    """Returns the name a reading is kept under, or raises Refused."""
    got = str(value or "")
    if not NAME.match(got):
        raise Refused("the name of a reading carries latin letters, digits and dashes and nothing else")
    return got


def _note(item):
    """Returns one note as the shelf keeps it, or raises Refused."""
    if not isinstance(item, dict):
        raise Refused("a note is an object")

    note_id = _line(item.get("id"), MAX_NOTE_ID)
    if not note_id:
        raise Refused("every note carries a name of its own")

    path = _line(item.get("path"))
    line = _count(item.get("line"))
    if not path or line <= 0:
        raise Refused("a note stands on a line of a file, and this one names none")

    text = _body(item.get("text"))
    if not text:
        raise Refused("a note with nothing written in it is not a note")
    if len(text) > MAX_TEXT:
        raise Refused(f"a note runs past {MAX_TEXT} characters")

    quote = _quote(item.get("quote"))
    if len(quote) > MAX_QUOTE:
        raise Refused(f"the line a note stands on runs past {MAX_QUOTE} characters")

    return {"id": note_id, "path": path, "line": line,
            "quote": quote, "text": text, "at": _stamp(item.get("at"))}


def _notes(raw):
    """Returns the notes of a reading, or raises Refused."""
    if raw is None:
        raw = []
    if not isinstance(raw, list):
        raise Refused("the notes of a reading are a list")
    if len(raw) > MAX_NOTES:
        raise Refused(f"a reading holds at most {MAX_NOTES} notes")

    out = []
    seen = set()
    for item in raw:
        note = _note(item)
        # The session finds a note again by its name. Two notes under one name
        # mean one of them is the note nobody can point at.
        if note["id"] in seen:
            raise Refused("every note carries a name of its own")
        seen.add(note["id"])
        out.append(note)
    return out


def clean(payload):
    """Returns the reading in the form the shelf keeps, or raises Refused."""
    if not isinstance(payload, dict):
        raise Refused("a reading is an object")

    review_id = name(payload.get("id"))

    session = _line(payload.get("session"), MAX_ID)
    if not session:
        raise Refused("a reading belongs to a conversation, and this one names none")

    return {
        "id": review_id,
        "session": session,
        "cwd": _line(payload.get("cwd")),
        "base": _line(payload.get("base")),
        "at": _stamp(payload.get("at")) or time.strftime(STAMP, time.gmtime()),
        "notes": _notes(payload.get("notes")),
    }


def card(review):
    """Returns what a list shows about a reading without carrying its notes."""
    notes = review.get("notes")
    return {
        "id": review.get("id", ""),
        "session": review.get("session", ""),
        "cwd": review.get("cwd", ""),
        "base": review.get("base", ""),
        "at": review.get("at", ""),
        "notes": len(notes) if isinstance(notes, list) else 0,
    }


class Shelf:
    """Readings on disk, one file each, named by the reading.

    One file each because that is what the session is handed: it opens the file
    it was pointed at and reads nothing else.
    """

    def __init__(self, path=None):
        self.path = path or STORE
        self.lock = threading.Lock()

    def _file(self, review_name):
        return os.path.join(self.path, review_name + ".json")

    def _read(self, review_name):
        try:
            with open(self._file(review_name), encoding="utf-8") as f:
                data = json.load(f)
        except (OSError, ValueError):
            return None
        return data if isinstance(data, dict) else None

    def put(self, review):
        """Writes a reading and returns where it lies, or why it was not written.

        A second write of the same reading is taken rather than refused: the
        panel only sends once, so a repeat is a send whose answer went missing,
        and the file it would write is the one already there.
        """
        with self.lock:
            try:
                os.makedirs(self.path, exist_ok=True)
            except OSError as e:
                return "", f"the shelf of readings is not writable: {e}"
            full = self._file(review["id"])
            tmp = full + ".new"
            try:
                with open(tmp, "w", encoding="utf-8") as f:
                    json.dump(review, f, ensure_ascii=False)
                os.replace(tmp, full)
            except OSError as e:
                try:
                    os.unlink(tmp)
                except OSError:
                    pass
                return "", f"the reading was not written: {e}"
        return full, ""

    def of(self, review_name):
        """Returns one reading whole, or None."""
        try:
            got = name(review_name)
        except Refused:
            # The name arrives from outside. A walk out of the shelf has to end
            # here rather than in a neighbouring directory.
            return None
        return self._read(got)

    def cards(self, session=None):
        """Returns the card of every reading, newest first."""
        try:
            names = os.listdir(self.path)
        except OSError:
            return []
        out = []
        for entry in names:
            if not entry.endswith(".json"):
                continue
            review = self._read(entry[: -len(".json")])
            if not review:
                continue
            if session and review.get("session") != session:
                continue
            out.append(card(review))
        out.sort(key=lambda c: str(c.get("at") or ""), reverse=True)
        return out

    def sweep(self):
        """Drops readings by age, by how long ago they went, by count and by room.

        A reading that has just been written is never among them: the largest
        one that can arrive is smaller than the room the shelf keeps, so the
        newest file cannot be the one the size rule reaches.
        """
        gone = []
        with self.lock:
            try:
                names = [n for n in os.listdir(self.path) if n.endswith(".json")]
            except OSError:
                return gone
            now = time.time()
            kept = []
            for entry in names:
                full = os.path.join(self.path, entry)
                try:
                    stat = os.stat(full)
                except OSError:
                    continue
                if now - stat.st_mtime > MAX_AGE:
                    gone.append(entry)
                    continue
                sent = seconds((self._read(entry[: -len(".json")]) or {}).get("at"))
                if sent and now - sent > SENT_AGE:
                    gone.append(entry)
                    continue
                kept.append((stat.st_mtime, stat.st_size, entry))
            kept.sort(reverse=True)
            room = 0
            for i, (_, size, entry) in enumerate(kept):
                room += size
                if i >= MAX_REVIEWS or room > MAX_STORE:
                    gone.append(entry)
            for entry in gone:
                try:
                    os.unlink(os.path.join(self.path, entry))
                except OSError:
                    pass
        return [n[: -len(".json")] for n in gone]


SHELF = Shelf()


def _recv(conn):
    """Reads one request whole: a reading does not fit in a single packet."""
    chunks = []
    size = 0
    while True:
        chunk = conn.recv(64 * 1024)
        if not chunk:
            break
        size += len(chunk)
        if size > MAX_REQUEST:
            raise Refused(f"the reading is longer than {MAX_REQUEST // (1024 * 1024)} MB")
        chunks.append(chunk)
    return b"".join(chunks)


def handle(conn, shelf=None):
    """Takes one reading and closes the connection."""
    shelf = shelf or SHELF
    with conn:
        try:
            conn.settimeout(20)
            payload = json.loads(_recv(conn).decode("utf-8"))
            review = clean(payload)
            where, why = shelf.put(review)
            if where:
                shelf.sweep()
                reply = {"ok": True, "id": review["id"], "path": where,
                         "notes": len(review["notes"])}
            else:
                reply = {"ok": False, "error": why}
        except Refused as e:
            reply = {"ok": False, "error": str(e)}
        except (OSError, ValueError, UnicodeDecodeError) as e:
            reply = {"ok": False, "error": f"the reading was not parsed: {e}"}
        try:
            conn.sendall(json.dumps(reply, ensure_ascii=False).encode("utf-8"))
        except OSError:
            pass


def listen():
    """Opens a 0600 socket in the shelf directory."""
    os.makedirs(SOCKET_DIR, exist_ok=True)
    path = os.path.join(SOCKET_DIR, SOCKET_NAME)
    try:
        os.unlink(path)
    except FileNotFoundError:
        pass
    sock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
    sock.bind(path)
    os.chmod(path, 0o600)
    sock.listen(8)
    return sock


def worker():
    """Runs the thread that takes the readings the panel sends."""
    try:
        sock = listen()
    except OSError as e:
        print(f"aacpanel-agent: readings will not reach the sessions, the socket did not come up: {e}", flush=True)
        return
    print(f"aacpanel-agent: listening for readings on {os.path.join(SOCKET_DIR, SOCKET_NAME)}", flush=True)
    while True:
        try:
            conn, _ = sock.accept()
        except OSError:
            return
        threading.Thread(target=handle, args=(conn,), daemon=True).start()

File: cmd/aacpanel/reviews.go
package main

import (
	"encoding/json"
	"errors"
	"io"
	"net/http"
	"strconv"
	"strings"
	"time"

	"aacpanel/internal/chat"
	"aacpanel/internal/store"
)

// What a reading is allowed to hold. The ceilings are here rather than in the
// database because a request is where a number too large arrives: the store
// would take it and the screen would then have to draw it.
const (
	maxNotes     = 200
	maxNoteText  = 4000
	maxNoteQuote = 2000
)

// apiReviews answers with the index: every reading with its status, newest
// first. A screen showing what is in hand reads this one answer instead of
// opening the readings one at a time.
func (s *Server) apiReviews(w http.ResponseWriter, r *http.Request) {
	if s.db == nil {
		http.Error(w, "readings are not kept: the database is not configured", http.StatusServiceUnavailable)
		return
	}
	limit, _ := strconv.Atoi(r.URL.Query().Get("limit"))
	list, err := s.db.Reviews(r.Context(), strings.TrimSpace(r.URL.Query().Get("session")), limit)
	if err != nil {
		http.Error(w, err.Error(), http.StatusServiceUnavailable)
		return
	}
	writeJSON(w, map[string]any{"reviews": list})
}

// apiReviewNew starts a reading and names it. The name is made here rather
// than in the browser: it becomes the name of a file on a shelf, and a name
// that arrives from outside is a path somebody else chose.
func (s *Server) apiReviewNew(w http.ResponseWriter, r *http.Request) {
	var body struct {
		Session string `json:"session"`
		Cwd     string `json:"cwd"`
		Base    string `json:"base"`
	}
	if err := json.NewDecoder(io.LimitReader(r.Body, 8<<10)).Decode(&body); err != nil {
		http.Error(w, "the reading did not parse: "+err.Error(), http.StatusBadRequest)
		return
	}
	session := strings.TrimSpace(body.Session)
	if session == "" {
		http.Error(w, "a reading belongs to a conversation, and this one names none", http.StatusBadRequest)
		return
	}
	if s.db == nil {
		http.Error(w, "readings are not kept: the database is not configured", http.StatusServiceUnavailable)
		return
	}

	id := reviewName(session, time.Now().UTC())
	review := store.Review{
		ID: id, Session: session,
		Cwd:  strings.TrimSpace(body.Cwd),
		Base: strings.TrimSpace(body.Base),
	}
	if err := s.db.SaveReview(r.Context(), review); err != nil {
		http.Error(w, err.Error(), http.StatusServiceUnavailable)
		return
	}
	made, err := s.db.ReviewOf(r.Context(), id)
	if err != nil {
		http.Error(w, err.Error(), http.StatusServiceUnavailable)
		return
	}
	writeJSON(w, made)
}

// reviewName is the name of a reading: the conversation it belongs to and when
// it started, so a shelf of them reads in order and says whose they are.
func reviewName(session string, at time.Time) string {
	safe := make([]rune, 0, len(session))
	for _, c := range strings.ToLower(session) {
		switch {
		case c >= 'a' && c <= 'z', c >= '0' && c <= '9':
			safe = append(safe, c)
		case c == '-' || c == '_':
			safe = append(safe, '-')
		}
	}
	name := strings.Trim(string(safe), "-")
	if name == "" {
		name = "reading"
	}
	if len(name) > 40 {
		name = name[:40]
	}
	return name + "-" + at.Format("20060102-150405")
}

// apiReview answers with one reading. A reading nobody has started comes back
// empty rather than missing: the screen opens it the same way either way.
func (s *Server) apiReview(w http.ResponseWriter, r *http.Request) {
	id, ok := reviewID(w, r)
	if !ok {
		return
	}
	if s.db == nil {
		http.Error(w, "readings are not kept: the database is not configured", http.StatusServiceUnavailable)
		return
	}
	review, err := s.db.ReviewOf(r.Context(), id)
	if err != nil {
		http.Error(w, err.Error(), http.StatusServiceUnavailable)
		return
	}
	writeJSON(w, review)
}

// apiReviewDraft writes down where a reading has got to.
func (s *Server) apiReviewDraft(w http.ResponseWriter, r *http.Request) {
	id, ok := reviewID(w, r)
	if !ok {
		return
	}
	var body struct {
		Session string             `json:"session"`
		Cwd     string             `json:"cwd"`
		Base    string             `json:"base"`
		Notes   []store.ReviewNote `json:"notes"`
	}
	if err := json.NewDecoder(io.LimitReader(r.Body, 1<<20)).Decode(&body); err != nil {
		http.Error(w, "the reading did not parse: "+err.Error(), http.StatusBadRequest)
		return
	}
	notes, err := cleanNotes(body.Notes)
	if err != nil {
		http.Error(w, err.Error(), http.StatusBadRequest)
		return
	}
	if strings.TrimSpace(body.Session) == "" {
		http.Error(w, "a reading belongs to a conversation, and this one names none", http.StatusBadRequest)
		return
	}
	if s.db == nil {
		http.Error(w, "the reading cannot be saved: the database is not configured", http.StatusServiceUnavailable)
		return
	}

	err = s.db.SaveReview(r.Context(), store.Review{
		ID:      id,
		Session: strings.TrimSpace(body.Session),
		Cwd:     strings.TrimSpace(body.Cwd),
		Base:    strings.TrimSpace(body.Base),
		Notes:   notes,
	})
	if errors.Is(err, store.ErrReviewSent) {
		http.Error(w, err.Error(), http.StatusConflict)
		return
	}
	if err != nil {
		http.Error(w, err.Error(), http.StatusServiceUnavailable)
		return
	}
	review, err := s.db.ReviewOf(r.Context(), id)
	if err != nil {
		http.Error(w, err.Error(), http.StatusServiceUnavailable)
		return
	}
	writeJSON(w, review)
}

// apiReviewFile writes the reading to the shelf and answers with the path of
// the file. Writing is a step of its own, before the signal: the file has to be
// there when the session comes to read it, and the session is told about it the
// way it is told anything else — through the queue of its composer.
func (s *Server) apiReviewFile(w http.ResponseWriter, r *http.Request) {
	id, ok := reviewID(w, r)
	if !ok {
		return
	}
	if s.db == nil {
		http.Error(w, "readings are not kept: the database is not configured", http.StatusServiceUnavailable)
		return
	}
	if !s.shelf.Available() {
		http.Error(w, chat.ErrNoShelf.Error(), http.StatusServiceUnavailable)
		return
	}

	review, err := s.db.ReviewOf(r.Context(), id)
	if err != nil {
		http.Error(w, err.Error(), http.StatusServiceUnavailable)
		return
	}
	if review.Session == "" {
		http.Error(w, "there is no such reading", http.StatusNotFound)
		return
	}
	if review.SentAt != nil {
		http.Error(w, store.ErrReviewSent.Error(), http.StatusConflict)
		return
	}
	if len(review.Notes) == 0 {
		http.Error(w, "a reading with no notes in it has nothing to send", http.StatusBadRequest)
		return
	}

	put := chat.ReviewPut{
		ID: review.ID, Session: review.Session, Cwd: review.Cwd, Base: review.Base,
		At:    time.Now().UTC().Format(time.RFC3339),
		Notes: make([]chat.ReviewNote, 0, len(review.Notes)),
	}
	for _, n := range review.Notes {
		put.Notes = append(put.Notes, chat.ReviewNote{
			ID: n.ID, Path: n.Path, Line: n.Line, Quote: n.Quote, Text: n.Text,
			At: n.At.UTC().Format(time.RFC3339),
		})
	}

	path, err := s.shelf.Put(r.Context(), put)
	if err != nil {
		if errors.Is(err, chat.ErrNoShelf) {
			http.Error(w, err.Error(), http.StatusServiceUnavailable)
			return
		}
		http.Error(w, err.Error(), http.StatusBadGateway)
		return
	}
	writeJSON(w, map[string]any{"id": review.ID, "path": path, "notes": len(review.Notes)})
}

// apiReviewSent records that a reading has gone to its session and where the
// file landed. It is a step of its own rather than part of writing the file:
// the signal travels to the session the way any message does, through the
// queue of a busy composer, and a reading is only settled once that has
// happened.
func (s *Server) apiReviewSent(w http.ResponseWriter, r *http.Request) {
	id, ok := reviewID(w, r)
	if !ok {
		return
	}
	var body struct {
		Path string `json:"path"`
	}
	if err := json.NewDecoder(io.LimitReader(r.Body, 8<<10)).Decode(&body); err != nil {
		http.Error(w, "the reply did not parse: "+err.Error(), http.StatusBadRequest)
		return
	}
	path := strings.TrimSpace(body.Path)
	if path == "" {
		http.Error(w, "a reading that has gone says where its file landed", http.StatusBadRequest)
		return
	}
	if s.db == nil {
		http.Error(w, "readings are not kept: the database is not configured", http.StatusServiceUnavailable)
		return
	}

	err := s.db.MarkReviewSent(r.Context(), id, path)
	if errors.Is(err, store.ErrReviewSent) {
		http.Error(w, err.Error(), http.StatusConflict)
		return
	}
	if err != nil {
		http.Error(w, err.Error(), http.StatusServiceUnavailable)
		return
	}
	review, err := s.db.ReviewOf(r.Context(), id)
	if err != nil {
		http.Error(w, err.Error(), http.StatusServiceUnavailable)
		return
	}
	writeJSON(w, review)
}

// apiReviewDrop removes a draft nobody sent.
func (s *Server) apiReviewDrop(w http.ResponseWriter, r *http.Request) {
	id, ok := reviewID(w, r)
	if !ok {
		return
	}
	if s.db == nil {
		http.Error(w, "readings are not kept: the database is not configured", http.StatusServiceUnavailable)
		return
	}
	err := s.db.DropReview(r.Context(), id)
	if errors.Is(err, store.ErrReviewSent) {
		http.Error(w, err.Error(), http.StatusConflict)
		return
	}
	if err != nil {
		http.Error(w, err.Error(), http.StatusServiceUnavailable)
		return
	}
	writeJSON(w, map[string]any{"dropped": id})
}

// reviewID is the name of a reading as it travels in a path: letters, digits
// and dashes. A name is made here and never taken from a path on disk — this
// is the one part of the viewer that writes, and what it writes is named by
// the panel, not by whoever asks.
func reviewID(w http.ResponseWriter, r *http.Request) (string, bool) {
	id := strings.TrimSpace(r.PathValue("id"))
	if id == "" || len(id) > 120 {
		http.Error(w, "that is not the name of a reading", http.StatusBadRequest)
		return "", false
	}
	for _, c := range id {
		ok := c == '-' || c == '_' ||
			(c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || (c >= '0' && c <= '9')
		if !ok {
			http.Error(w, "the name of a reading carries letters, digits and dashes and nothing else",
				http.StatusBadRequest)
			return "", false
		}
	}
	return id, true
}

// cleanNotes holds the notes to what a screen can draw and a file can carry.
// The quote travels as it came, an empty line included: it is how a note is
// found again once the branch moves, and trimming it would make a note on an
// empty line indistinguishable from a note that lost its place.
func cleanNotes(raw []store.ReviewNote) ([]store.ReviewNote, error) {
	out := make([]store.ReviewNote, 0, len(raw))
	if len(raw) > maxNotes {
		return nil, errors.New("a reading holds at most " + strconv.Itoa(maxNotes) + " notes")
	}
	seen := map[string]bool{}
	for _, n := range raw {
		n.ID = strings.TrimSpace(n.ID)
		n.Path = strings.TrimSpace(n.Path)
		n.Text = strings.TrimSpace(n.Text)
		n.Quote = strings.TrimRight(n.Quote, "\r\n")
		if n.ID == "" || seen[n.ID] {
			return nil, errors.New("every note carries a name of its own")
		}
		seen[n.ID] = true
		if n.Path == "" || n.Line <= 0 {
			return nil, errors.New("a note stands on a line of a file, and this one names none")
		}
		if n.Text == "" {
			return nil, errors.New("a note with nothing written in it is not a note")
		}
		if len(n.Text) > maxNoteText {
			n.Text = n.Text[:maxNoteText]
		}
		if len(n.Quote) > maxNoteQuote {
			n.Quote = n.Quote[:maxNoteQuote]
		}
		if n.At.IsZero() {
			n.At = time.Now().UTC()
		}
		out = append(out, n)
	}
	return out, nil
}
